### DISCUSSION TITLE
parfor and optimizer encounters the error

### DISCUSSION THREAD

**Discussion by arielj-weibo** [yalmip/YALMIP#1116] [General discussion]
**parfor s=1:1000
    b=sum(sum(W([m*s-m+1:m*s],:).*xod))+rho.*sum(sum((xod-X_mean).*(xod-X_mean)));
    z=-a+b;
    % solver setting
    ops = sdpsettings('verbose',0,'solver','cplex');
    p=optimizer(C,z,ops,zods,xod);
    xods{s}=p(zod(:,s))
end**

**Comment by johanlofberg** [comment]
Which version of MATLAB are you using? Old version had poor functionality in parfor

**Reply by johanlofberg** [reply]
Your code makes no sense though as you create a optimizer object inside a loop and only use it once. It will only be slower than simply calling optimize for specific data and model once. You use optimizer to construct an optimizer object outside a loop, and then call it for changing data in the loop.

**Reply by arielj-weibo** [reply]
2020a

**Reply by arielj-weibo** [reply]
Actually, I have tested that if I don't use the optimizer, just use **_for_**, it is much slower. Thus, optimizer is quicker. However,I want to be more quicker so i think of the **_parfor_**

**Reply by johanlofberg** [reply]
You would have to supply a reproducible example, because that should be impossible

**Reply by arielj-weibo** [reply]
I want to solve 1000 scenarios's model with only **the objective function b** and one variable **zod** in the constraints is about **s**, and the b is associated with **matrix W** which is also associated with s.

**Reply by johanlofberg** [reply]
It is still not reproducible (you use local data)

**Reply by arielj-weibo** [reply]
I sent it to your email，which is much more convenient for your watching

**Reply by johanlofberg** [reply]
Very weird code where you have some stuff treated via the intended use of an optimizer, and then some parts being treated by redefining it.

The standard use would be

```
ops = sdpsettings('verbose',1,'solver','cplex');
Wpart= sdpvar(same size as xod);
z=  -a + sum(sum(Wpart.*xod))+rho.*sum(sum((xod-X_mean).*(xod-X_mean)));
p=optimizer(C,z,ops,{Wpart,zods},xod);

for s=1:1000
Wpart = ...
    xods{s}=p({Wpart,zod(:,s)});
end
```

**Reply by arielj-weibo** [reply]
Thank you so much for your reply！

**Reply by arielj-weibo** [reply]
Can I use parfor with the optimizer? just like the following.I just transfer the _for_ to the _parfor_

ops = sdpsettings('verbose',1,'solver','cplex');
Wpart= sdpvar(same size as xod);
z=  -a + sum(sum(Wpart.*xod))+rho.*sum(sum((xod-X_mean).*(xod-X_mean)));
p=optimizer(C,z,ops,{Wpart,zods},xod);

parfor s=1:1000
Wpart = ...
    xods{s}=p({Wpart,zod(:,s)});
end

**Reply by arielj-weibo** [reply]
Each time i use the change, the matlab will encounter the error.
I really want to speed up the solution.
And i also come up with the thread setting of parfor, just like:
**_ops.cplex.Param.threads.Cur=1_**

**Reply by johanlofberg** [reply]
No idea what you are asking

**Reply by arielj-weibo** [reply]
Can the optimizer be used in the parfor loop like this? 


z= -a + sum(sum(Wpart.*xod))+rho.sum(sum((xod-X_mean).(xod-X_mean)));
p=optimizer(C,z,ops,{Wpart,zods},xod);

**parfor** s=1:1000
      Wpart = ...
      xods{s}=p({Wpart,zod(:,s)});
end

**Reply by johanlofberg** [reply]
Yes

```
>> sdpvar a x
P = optimizer([-1 <= x <= 1],(x-a)^2,[],a,x);
Q = [];
data = -2:.25:2;
parfor i = 1:length(data)
    Q = [Q P(data(i))];
end
Q

Starting parallel pool (parpool) using the 'local' profile ...
Connected to the parallel pool (number of workers: 2).


Q =

  Columns 1 through 6

   -1.0000   -1.0000   -1.0000   -1.0000   -0.9998   -0.7500

  Columns 7 through 12

   -0.5000   -0.2500         0    0.2500    0.5000    0.7500

  Columns 13 through 17

    0.9998    1.0000    1.0000    1.0000    1.0000
```

**Reply by arielj-weibo** [reply]
thank you so much and i will try it again!
